### PR TITLE
Use bytecode deployer for ICA auth

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Replaces `OwnableMulticall` (`owner` in storage) with `DeployerMulticall` (`deployer` in bytecode) for optimizing ICA deployment costs

### Backward compatibility

No, bytecode changes implies new ICA addresses

### Testing

Unit Tests

Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1969

### CODE REVIEW

**Review by asaj — DISMISSED** [review]
Clever. Can we delete OwnableMulticall?

File: typescript/sdk/src/middleware/deploy.ts [hunk 1]
```diff
@@ -60,12 +64,16 @@ export abstract class MiddlewareRouterDeployer<
...
+    chain: ChainName,
+    config: MiddlewareRouterConfig,
+  ): Promise<any> {
+    const signer = await this.multiProvider.getSignerAddress(chain);
     return [
       config.mailbox,
       config.interchainGasPaymaster,
       config.interchainSecurityModule ?? ethers.constants.AddressZero,
-      config.owner,
+      signer, // set signer as owner for subsequent initialization calls
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/middleware/deploy.ts:76]: It looks like we do so in HyperlaneRouterDEployer, please capture this in a comment here

File: typescript/sdk/src/middleware/deploy.ts [hunk 2]
```diff
@@ -112,9 +121,68 @@ export class InterchainAccountDeployer extends MiddlewareRouterDeployer<
...
   ) {
-    super(multiProvider, configMap, interchainAccountFactories, create2salt);
+    super(multiProvider, configMap, interchainAccountFactories, create2Salt);
+  }
+
+  async deployContracts(
+    chain: ChainName,
+    config: InterchainAccountConfig,
+  ): Promise<InterchainAccountContracts> {
+    // adapted from super.deployProxy to use a dummy implementation
```
Comment (asaj) [inline typescript/sdk/src/middleware/deploy.ts:133]: This is all weird enough that it deserves a longer comment explaining what were doing and especially *why* we're doing it

**Review by asaj — APPROVED** [review]
Sweet

File: solidity/test/InterchainAccountRouter.t.sol [hunk 3]
```diff
@@ -152,6 +152,8 @@ contract InterchainAccountRouterTest is Test {
         // receive value after deployed
         remoteRouter.getDeployedInterchainAccount(originDomain, address(this));
         assert(ica.code.length > 0);
+        // necessary in forge tests for receive() to be executed upon transfer ?
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/test/InterchainAccountRouter.t.sol:155]: Actually please do not merge until we resolve this question

File: typescript/sdk/src/middleware/deploy.ts [hunk 4]
```diff
@@ -60,12 +64,16 @@ export abstract class MiddlewareRouterDeployer<
     return [] as any;
   }
 
-  initializeArgs(config: MiddlewareRouterConfig): any {
+  async initializeArgs(
+    chain: ChainName,
+    config: MiddlewareRouterConfig,
+  ): Promise<any> {
```
Comment (jmrossy) [inline typescript/sdk/src/middleware/deploy.ts:70]: Optional: fix type here? Not sure why we cast return type to any

File: typescript/sdk/src/middleware/deploy.ts [hunk 5]
```diff
@@ -112,9 +121,68 @@ export class InterchainAccountDeployer extends MiddlewareRouterDeployer<
...
+        implementation: implementation.address,
+        proxy: proxy.address,
+      },
+    );
+
+    return {
+      proxyAdmin,
+      proxiedRouter,
+      router: proxiedRouter.contract, // for backwards compatibility
+    } as any;
```
Comment (jmrossy) [inline typescript/sdk/src/middleware/deploy.ts:185]: ditto

**Review by jmrossy — APPROVED** [review]

**Pull request merged**